feature-utils/silly-i18n/rollup-plugin.cjs

```diff
@@ -0,0 +1,57 @@
+const fs = require("fs");
+const path = require("path");
+const util = require("util");
+const glob = require("glob");
+
+const DIRNAME = "locales";
+const FILEEXT = ".json";
+const PLUGIN = "silly-i18n";
+const SUFFIX = `?${PLUGIN}`;
+
+exports.default = function (options) {
+    return {
+        name: PLUGIN,
+
+        async resolveId(source, importer) {
+            if (source == `!${PLUGIN}`) {
+                for (let dir = importer; dir != (dir = path.dirname(dir)); ) {
+                    const subdir = path.join(dir, DIRNAME);
+                    const exists = await fs.promises.stat(subdir).then(
+                        (stat) => stat.isDirectory(),
+                        () => false
+                    );
+                    if (exists) return `${subdir}${SUFFIX}`;
+                }
+                throw new Error(`"${DIRNAME}" directory not found`);
```

Comment:
OK, let me ellaborate here. Above you're not searching for `DIRNAME` but for `dir/DIRNAME` (not entirely clear about this, though). You're going over a loop, checking different combinations; if the loop arrives here it's not because DIRNAME does not exist, it's because every one of the different `dir/DIRNAME` don't. So the main issue here is not so much that there should be a single check for DIRNAME and exit early (but I'll need to see what `importer` iterates over) but for the fact that somehow, a message saying what's been tested and does not exist would possibly be better.

Reply:
We are traversing up the file hierarchy from the module that invokes the import until we find a directory that has a `locales` sub-directory.

Reply:
Again, this is additional complexity caused by the fact that you're generating the object *every single time* it finds the pattern. And you're doing so using heuristics for the location of the `locales` file, as well as name; these are not specced anywhere.
Again, a static generator, run every time a new `locales` file is added, would simply get as argument the location for this directory, and generate a single file, that would be included in the repo. Have you considered that option? That would add 0 time to the build process, 0 refactoring would be needed, since the same path would be reused, and in general modification of a single file (the generator) would be needed. Is there something I'm missing here, that makes the rollup plugin superior?

Reply:
The JSON files are read, parsed and the object is generated only once. But yeah, the `locales` directory is located for every import. This is not any different for package imports where Rollup traverses the file system in a similar fashion to find the `node_modules` folder and the package in there. So I wouldn't be concerned about build speed here.

So you suggest a standalone script that you have to run manually after changing any translations and add the generated code to the repository? I don't know about that. It's generally good practice to not add generated code to the repository. The generates code can get out of sync, bloats the repository, adds more complexity to the development process as people need to know about it and remember to run it manually.